cpp/src/arrow/util/decimal.cc

```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
   return ::arrow::bit_util::FromBigEndian(result);
 }
 
+Result<Decimal32> Decimal32::FromBigEndian(const uint8_t* bytes, int32_t length) {
+  static constexpr int32_t kMinDecimalBytes = 1;
+  static constexpr int32_t kMaxDecimalBytes = 4;
+
+  if (ARROW_PREDICT_FALSE(length < kMinDecimalBytes || length > kMaxDecimalBytes)) {
+    return Status::Invalid("Length of byte array passed to Decimal32::FromBigEndian was ",
+                           length, ", but must be between ", kMinDecimalBytes, " and ",
+                           kMaxDecimalBytes);
+  }
+
+  const bool is_negative = static_cast<int8_t>(bytes[0]) < 0;
+
+  uint32_t result = 0;
+  memcpy(reinterpret_cast<uint8_t*>(&result) + kMaxDecimalBytes - length, bytes, length);
+
+  const auto value = bit_util::FromBigEndian(result);
+  const int32_t bits_offset = std::max(0, length - kMaxDecimalBytes);
+  if (bits_offset == 8) {
+    return Decimal32(value);
+  }
+
+  int32_t final_value = -1 * (is_negative && length < kMaxDecimalBytes);
+  final_value = SafeLeftShift(final_value, bits_offset * CHAR_BIT);
+  final_value |= value;
```

Comment:
This looks... complicated?

Reply:
I adapted this from the `Decimal128::FromBigEndian` but now that I'm re-reading it, am I wrong to think that I could just do `FromBigEndian` on the int32_t? (not `uint32_t`)

Reply:
Ah, I can't just blindly do `FromBigEndian` because i'm not guaranteed that `length` is exactly 4 bytes, hence why this shifting logic is needed for negative values.